APE_Bayesian_modeling.rst

```diff
@@ -0,0 +1,228 @@
+  and returns the prior evaluated at those parameters.
+
+The implementation of the first two ways will be done by having a generic ``prior``
+function similar to the third way that takes all the parameters as input.
+This generic ``prior`` function then uses any ``Parameter.prior`` functions and then
+any ``Parameter.bounds`` implied flat prior functions.  For any ``Parameters``
+without any prior, this function would just give a constant prior (e.g., 1).
+
+
+.. code-block:: python
+
+    class Parameter:
+        ...
+        self._prior = <user_supplied_prior>
+
+        @property
+        def prior(self):
+            if self._prior is not None:
+                return self._prior
+            elif (not any(b is None for b in self.bounds)) and
+                self.value > self.min and self.value < self.max:
+                return 1.0
+            else:
+                return 0.0
+
+
+    class Model:
+        ...
+        def prior(self):
+            """
+            Returns the combined priors of all parameters.
+
+            Evaluates all priors and returns their product.
+            """
+
+
+Statistics
+==========
+
+A statistic that combines the likelihood that the data fits the model with the
+prior function will be implemented.  This is called the posterior and is the
+multiplication of the likelihood function by the prior function.  The maximum
+of the posterior provides the best fit model.  For numerical reasons, the
+natural log of the likelihood and prior function are added and the posterior
+probability is reported as ln(prob).
+
+The current set of fitters are based on optimizers, which minimize the statistic.
+Currently, the only implemented statistic is the Gaussian chisqr.  At a minimum,
+a new statistic that combines a Gaussian probability (i.e., exp(-chisqr/2)) with
+the prior function is needed.  An alternative would be to create a new class
+(`Posterior`?) that combines both a statistic (i.e., likelilhood) and prior
+allowing the function to be maximized.
+
+Samplers
+========
+
+A new module akin to fitting needs to be added.  This module would be based
+around a ``Sampler`` class (similar to the ``Optimization`` class), which would
+provide the generic interface for different samplers.  Unlike the existing
+Optimizers where the best fitting model is the result, Samplers provide a
+sampling of the posterior where the number of samples is set as an input.
+Effectively, the samplers provide a set of models where the set reflects the
+posterior.  This set of models is usually returned as a 1D array of values
+for each model parameter (called a chain), where the posterior is reflected
+as the frequency of a model parameter in the chain or as a parallel 1D array
+of posterior weights. The chain for each parameter will be stored in
+`Parameter.sampled_posterior` as an ``astropy.uncertainty`` distribution.
+
+At least two samplers will be implemented.  The specific samplers planned are
+`emcee` and `dynasty`.
```

Comment:
I'm not convinced those two implementations should live in astropy itself. There are both practical and philosophical arguments to keep in mind.

- If they live in astropy we have to test them. That means adding emcee and dynasty to our test dependencies, potentially breaking tests when those packages upgrade etc. That's a lot of work for something that's not core astronomy specific (fitting is done in many fields of science).
- Should astropy really pick out and endorse those two packages? Why not BaysicFitting or Sherpa or Ultranest?
- We have precedence. [Sherpa](https://sherpa.readthedocs.io) also provides substantial extensions to `astropy.modelling`; it has many more models than Astropy has (a large fraction, but not all of those are X-ray specific), about a dozen different statistics, half a dozen optimizes (beyond those already part of scipy), and also a MCMC sampler. A few years ago, we build [saba](https://saba.readthedocs.io) as an astropy GSOC project to bridge astropy and sherpa through a translation layer. (Disclosure: I'm supposed to maintain that package and have not done a good job, so it's not currently compatible with astropy 5.0.)

So, I think we should consider just defining the generic API in astropy (and test with a dummy) and implement the actual bridging code either in a separate package or contribute it to e.g. dynasty.

Reply:
Ah.  The APE should say that *interfaces* to at least two samplers should be implemented.

Reply:
So, fully agree with your comment!  Interfaces to more than just the names two packages would be great as you clearly point out there are many other options.

Reply:
Your comment about ``saba`` and maintenance to keep it current with ``astropy`` is one of the considerations between interfacing with an external package (e.g., ``BayesicFitting``) and picking the route outlined in this APE.

Reply:
It's definitely worth considering both ways. But I recall the discussion in https://github.com/astropy/astropy/issues/12448 where there is not even agreement to make scipy a dependency unless we also provide ways to test and run without scipy. Now, that issue was about making it a required, not just an optional dependency, but scipy is also way more established (number of contributors, testing on unusual systems, packaging, ...) than any of the fitting packages we discuss here.

(I'm commenting so extensively here, because my wife does not feel well and I might have to pick up our son from school, which mean I would miss our live zoom discussion later today. So, no need to reply to every one of my comments, just consider them in case I'm not there later.)

Reply:
But replying to your comments is more fun than what I'm supposed to be doing.  ;-)

Seriously, all the external optimizers and samplers used will have to be optional dependencies.  What do do if they aren't installed is a good question.  Maybe that argues for a single optimizer and a single sampler to be implemented in astropy?  Maybe the simplest one possible?